File: .claude/hooks/log-cli-tools.php
#!/usr/bin/env php
<?php
/**
 * PostToolUse hook: Log Codex/Gemini CLI input/output to JSONL file.
 *
 * Triggers after Bash tool calls containing 'codex' or 'gemini' commands.
 * Logs are stored in .claude/logs/cli-tools.jsonl
 *
 * All agents (Claude Code, subagents, Codex, Gemini) can read this log.
 */

define('LOG_DIR', dirname(__FILE__) . '/../logs');
define('LOG_FILE', LOG_DIR . '/cli-tools.jsonl');

function extractCodexPrompt(string $command): ?string
{
    // Pattern: codex exec ... "prompt" or codex exec ... 'prompt'
    $patterns = [
        '/codex\s+exec\s+.*?--full-auto\s+"([^"]+)"/s',
        "/codex\s+exec\s+.*?--full-auto\s+'([^']+)'/s",
        '/codex\s+exec\s+.*?"([^"]+)"\s*2>\/dev\/null/s',
        "/codex\s+exec\s+.*?'([^']+)'\s*2>\/dev\/null/s",
    ];
    foreach ($patterns as $pattern) {
        if (preg_match($pattern, $command, $matches)) {
            return trim($matches[1]);
        }
    }
    return null;
}

function extractGeminiPrompt(string $command): ?string
{
    // Pattern: gemini -p "prompt" or gemini -p 'prompt'
    $patterns = [
        '/gemini\s+-p\s+"([^"]+)"/s',
        "/gemini\s+-p\s+'([^']+)'/s",
    ];
    foreach ($patterns as $pattern) {
        if (preg_match($pattern, $command, $matches)) {
            return trim($matches[1]);
        }
    }
    return null;
}

function extractModel(string $command): ?string
{
    if (preg_match('/--model\s+(\S+)/', $command, $matches)) {
        return $matches[1];
    }
    return null;
}

function truncateText(string $text, int $maxLength = 2000): string
{
    if (mb_strlen($text) <= $maxLength) {
        return $text;
    }
    $totalChars = mb_strlen($text);
    return mb_substr($text, 0, $maxLength) . "... [truncated, {$totalChars} total chars]";
}

function logEntry(array $entry): void
{
    if (!is_dir(LOG_DIR)) {
        mkdir(LOG_DIR, 0755, true);
    }
    file_put_contents(
        LOG_FILE,
        json_encode($entry, JSON_UNESCAPED_UNICODE) . "\n",
        FILE_APPEND | LOCK_EX
    );
}

function main(): void
{
    // Read hook input from stdin
    $rawInput = file_get_contents('php://stdin');
    $hookInput = json_decode($rawInput, true);
    if ($hookInput === null) {
        return;
    }

    // Only process Bash tool calls
    $toolName = $hookInput['tool_name'] ?? '';
    if ($toolName !== 'Bash') {
        return;
    }

    // Get command and output
    $toolInput = $hookInput['tool_input'] ?? [];
    $toolResponse = $hookInput['tool_response'] ?? [];

    $command = $toolInput['command'] ?? '';
    $output = $toolResponse['stdout'] ?? $toolResponse['content'] ?? '';

    // Check if this is a codex or gemini command
    $commandLower = strtolower($command);
    $isCodex = str_contains($commandLower, 'codex');
    $isGemini = str_contains($commandLower, 'gemini') && !str_contains($commandLower, 'codex');

    if (!$isCodex && !$isGemini) {
        return;
    }

    // Extract prompt based on tool type
    if ($isCodex) {
        $tool = 'codex';
        $prompt = extractCodexPrompt($command);
        $model = extractModel($command) ?? 'gpt-5.2-codex';
    } else {
        $tool = 'gemini';
        $prompt = extractGeminiPrompt($command);
        $model = 'gemini-3-pro-preview';
    }

    if ($prompt === null) {
        // Could not extract prompt, skip logging
        return;
    }

    // Determine success
    $exitCode = $toolResponse['exit_code'] ?? 0;
    $success = $exitCode === 0 && !empty($output);

    // Create log entry
    $entry = [
        'timestamp' => gmdate('c'),
        'tool' => $tool,
        'model' => $model,
        'prompt' => truncateText($prompt),
        'response' => $output ? truncateText($output) : '',
        'success' => $success,
        'exit_code' => $exitCode,
    ];

    logEntry($entry);

    // Output notification (shown to user via hook output)
    echo json_encode([
        'result' => 'continue',
        'message' => "[LOG] " . ucfirst($tool) . " call logged to .claude/logs/cli-tools.jsonl",
    ]) . "\n";
}

main();

File: .agents/hooks/log-cli-tools.py
#!/usr/bin/env python3
"""
PostToolUse hook: Log Codex CLI input/output to JSONL file.

Triggers after Bash tool calls containing 'codex' commands.
Logs are stored in .agents/logs/cli-tools.jsonl

All agents (Claude Code, subagents, Codex) can read this log.
"""

import json
import re
import sys
from datetime import UTC, datetime
from pathlib import Path

LOG_DIR = Path(__file__).parent.parent / "logs"
LOG_FILE = LOG_DIR / "cli-tools.jsonl"


def extract_codex_prompt(command: str) -> str | None:
    """Extract prompt from codex exec command."""
    # Pattern: codex exec ... "prompt" or codex exec ... 'prompt'
    patterns = [
        r'codex\s+exec\s+.*?"([^"]+)"\s*2>/dev/null',
        r"codex\s+exec\s+.*?'([^']+)'\s*2>/dev/null",
    ]
    for pattern in patterns:
        match = re.search(pattern, command, re.DOTALL)
        if match:
            return match.group(1).strip()
    return None


def extract_model(command: str) -> str | None:
    """Extract model name from command."""
    match = re.search(r"--model\s+(\S+)", command)
    return match.group(1) if match else None


def truncate_text(text: str, max_length: int = 2000) -> str:
    """Truncate text if too long."""
    if len(text) <= max_length:
        return text
    return text[:max_length] + f"... [truncated, {len(text)} total chars]"


def log_entry(entry: dict) -> None:
    """Append entry to JSONL log file."""
    LOG_DIR.mkdir(parents=True, exist_ok=True)
    with open(LOG_FILE, "a", encoding="utf-8") as f:
        f.write(json.dumps(entry, ensure_ascii=False) + "\n")


def process_bash(data: dict) -> str | None:
    """Log a Codex CLI call from a Bash tool invocation, if present.

    Pure-ish function (its side effect is the JSONL append, which is the
    whole point of this hook) so the post-bash-check.py dispatcher can
    call it in-process. Returns a confirmation context string, or None if
    there was nothing to log. Standalone main() below still reads stdin
    directly for backwards-compatible direct invocation.
    """
    # Only process Bash tool calls
    tool_name = data.get("tool_name", "")
    if tool_name != "Bash":
        return None

    # Get command and output
    tool_input = data.get("tool_input", {})
    tool_response = data.get("tool_response", {})

    command = tool_input.get("command", "")
    output = tool_response.get("stdout", "") or tool_response.get("content", "")

    # Check if this is a codex command
    if "codex" not in command.lower():
        return None

    prompt = extract_codex_prompt(command)
    model = extract_model(command) or "gpt-5.6-sol"

    if not prompt:
        # Could not extract prompt, skip logging
        return None

    # Determine success
    exit_code = tool_response.get("exit_code", 0)
    success = exit_code == 0 and bool(output)

    # Create log entry
    entry = {
        "timestamp": datetime.now(UTC).isoformat(),
        "tool": "codex",
        "model": model,
        "prompt": truncate_text(prompt),
        "response": truncate_text(output) if output else "",
        "success": success,
        "exit_code": exit_code,
    }

    log_entry(entry)

    return "[LOG] Codex call logged to .agents/logs/cli-tools.jsonl"


def main() -> None:
    # Read hook input from stdin. This also handles the TaskCompleted
    # wiring: TaskCompleted payloads have no tool_name == "Bash", so
    # process_bash() returns None and this hook is a no-op for them.
    try:
        hook_input = json.load(sys.stdin)
    except json.JSONDecodeError:
        return

    context = process_bash(hook_input)
    if context is None:
        return

    # Output notification via hookSpecificOutput
    print(
        json.dumps(
            {
                "hookSpecificOutput": {
                    "additionalContext": context,
                }
            }
        )
    )


if __name__ == "__main__":
    main()
